compiler/rustc_borrowck/src/region_infer/mod.rs

```diff
@@ -2151,22 +2150,19 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 });
 
                 if let Some(field) = field {
-                    categorized_path[i].category =
+                    let mut blame_vec: Vec<_> = categorized_path.into_iter().collect();
+                    blame_vec[i].category =
                         ConstraintCategory::Return(ReturnConstraint::ClosureUpvar(field));
+                    let updated_categorized_path: indexmap::IndexSet<_> =
+                        blame_vec.into_iter().collect();
+                    categorized_path = updated_categorized_path
                 }
             }
 
             return categorized_path[i].clone();
         }
 
-        // If that search fails, that is.. unusual. Maybe everything
-        // is in the same SCC or something. In that case, find what
-        // appears to be the most interesting point to report to the
-        // user via an even more ad-hoc guess.
-        categorized_path.sort_by(|p0, p1| p0.category.cmp(&p1.category));
-        debug!("best_blame_constraint: sorted_path={:#?}", categorized_path);
-
-        categorized_path.remove(0)
```

Comment:
Might be able to replace this logic via this fixme: https://github.com/rust-lang/rust/blob/f3f865e6a15bc5a580572f9b836913b2a39737dd/compiler/rustc_borrowck/src/diagnostics/region_errors.rs#L401-L402